sdk/core/azure-core/test/ut/curl_connection_pool.cpp

```diff
@@ -0,0 +1,172 @@
+        // And the connection key for the connection we got is the expected
+        EXPECT_EQ(connection->GetConnectionKey(), expectedConnectionKey);
+
+        auto session = std::make_unique<Azure::Core::Http::CurlSession>(
+            req, std::move(connection), options.HttpKeepAlive);
+        // Simulate connection was used already
+        session->m_lastStatusCode = Azure::Core::Http::HttpStatusCode::Ok;
+        session->m_sessionState = Azure::Core::Http::CurlSession::SessionState::STREAMING;
+      }
+      // Check that after the connection is gone, it is moved back to the pool
+      EXPECT_EQ(Azure::Core::Http::CurlConnectionPool::ConnectionPoolIndex.size(), 1);
+      auto values = Azure::Core::Http::CurlConnectionPool::ConnectionPoolIndex.begin();
+      EXPECT_EQ(values->second.size(), 1);
+      EXPECT_EQ(values->second.begin()->get()->GetConnectionKey(), expectedConnectionKey);
+
+      // Now test that using a different connection config won't re-use the same connection
+      std::string const CAinfo = "someFakePath";
+      std::string const secondExpectedKey = "httpbin.org" + CAinfo + "011";
+      {
+        // Creating a new connection with default options
+        Azure::Core::Http::CurlTransportOptions options;
+        options.CAInfo = CAinfo;
+        auto connection = Azure::Core::Http::CurlConnectionPool::GetCurlConnection(req, options);
+        EXPECT_EQ(connection->GetConnectionKey(), secondExpectedKey);
+        // One connection still in the pool after getting a new connection and with first expected
+        // key
+        EXPECT_EQ(Azure::Core::Http::CurlConnectionPool::ConnectionPoolIndex.size(), 1);
+        EXPECT_EQ(
+            Azure::Core::Http::CurlConnectionPool::ConnectionPoolIndex.begin()
+                ->second.begin()
+                ->get()
+                ->GetConnectionKey(),
+            expectedConnectionKey);
+
+        auto session = std::make_unique<Azure::Core::Http::CurlSession>(
+            req, std::move(connection), options.HttpKeepAlive);
+        // Simulate connection was used already
+        session->m_lastStatusCode = Azure::Core::Http::HttpStatusCode::Ok;
+        session->m_sessionState = Azure::Core::Http::CurlSession::SessionState::STREAMING;
+      }
+
+      // Now there should be 2 index wit one connection each
+      EXPECT_EQ(Azure::Core::Http::CurlConnectionPool::ConnectionPoolIndex.size(), 2);
+      values = Azure::Core::Http::CurlConnectionPool::ConnectionPoolIndex.begin();
+      EXPECT_EQ(values->second.size(), 1);
+      EXPECT_EQ(values->second.begin()->get()->GetConnectionKey(), expectedConnectionKey);
+      values++;
+      EXPECT_EQ(values->second.size(), 1);
+      EXPECT_EQ(values->second.begin()->get()->GetConnectionKey(), secondExpectedKey);
+
+      // Test re-using same custom config
+      {
+        // Creating a new connection with default options
+        Azure::Core::Http::CurlTransportOptions options;
+        options.CAInfo = CAinfo;
```

Comment:
We don't have high branch coverage here, because we should also add a test where proxy is non-empty, and one where httpkeepalive is set, so that all the if(true) 1, else 0 branches are covered in the connection key generation.

Reply:
The number of tests is 2^n, where n is the number of options.
A few combinations should be more than enough.
Otherwise, I would expend more time creating a test generator, but doing it manually would be crazy!:)